dbms/src/TiDB/Schema/SchemaBuilder.cpp

```diff
@@ -1439,23 +1453,122 @@ void SchemaBuilder<Getter, NameMapper>::syncAllSchema()
 
     /// Drop all unmapped databases
     const auto & dbs = context.getDatabases();
-    for (auto it = dbs.begin(); it != dbs.end(); it++)
+    for (const auto & [db_name, db_ptr] : dbs)
     {
-        auto db_keyspace_id = SchemaNameMapper::getMappedNameKeyspaceID(it->first);
-        if (db_keyspace_id != keyspace_id)
-        {
+        std::ignore = db_ptr;
+        // ignore the database that not belongs to this keyspace
+        if (auto db_keyspace_id = SchemaNameMapper::getMappedNameKeyspaceID(db_name); db_keyspace_id != keyspace_id)
             continue;
-        }
-        if (!latest_db_nameset.nonThreadSafeContains(it->first) && !isReservedDatabase(context, it->first))
+
+        if (!latest_db_nameset.nonThreadSafeContains(db_name) && !isReservedDatabase(context, db_name))
         {
-            applyDropDatabaseByName(it->first);
-            LOG_INFO(log, "Database {} dropped during sync all schemas", it->first);
+            applyDropDatabaseByName(db_name);
+            LOG_INFO(log, "Database {} dropped during sync all schemas", db_name);
         }
     }
 
     LOG_INFO(log, "Sync all schemas end");
 }
 
+
+template <typename Getter, typename NameMapper>
+void SchemaBuilder<Getter, NameMapper>::tryFixPartitionsBelongingDatabase()
+{
+    size_t num_renamed = 0;
+    auto part_to_db_id = table_id_map.getAllPartitionsBelongDatabase();
+    for (const auto & [db_name, db_ptr] : context.getDatabases())
+    {
+        if (part_to_db_id.empty())
+            return;
+
+        if (db_name == "system")
+            continue;
+        // ignore the database that not belongs to this keyspace
+        if (auto db_keyspace_id = SchemaNameMapper::getMappedNameKeyspaceID(db_name); db_keyspace_id != keyspace_id)
+            continue;
+
+        // Get the `database_id` parsed from local disk "IDatabase" name
+        const DatabaseID database_id = SchemaNameMapper::tryGetDatabaseID(db_name).value_or(-1);
+        if (database_id == -1)
+        {
+            LOG_WARNING(
+                log,
+                "FixPartitionsDatabase: fail to parse database_id from database_name, ignore, db_name={}",
+                db_name);
+            continue;
+        }
+
+        for (auto tbl_iter = db_ptr->getIterator(context); tbl_iter->isValid(); tbl_iter->next())
+        {
+            const auto table_name = tbl_iter->table()->getTableName();
+            auto opt_tbl_id = SchemaNameMapper::tryGetTableID(table_name);
+            if (!opt_tbl_id)
+            {
+                LOG_WARNING(
```

Comment:
In what case could this happen?

Reply:
Cases under "tests/delta-merge-test" may create table with the given table_name rather that "t_${table_id}"